dipy/align/imaffine.py

```diff
@@ -0,0 +1,777 @@
+import numpy as np
+import numpy.linalg as npl
+import scipy.ndimage as ndimage
+from dipy.align import floating
+import dipy.align.vector_fields as vf
+from dipy.align.mattes import MattesBase, sample_domain_regular
+from dipy.core.optimize import Optimizer
+from dipy.align.imwarp import (get_direction_and_spacings, ScaleSpace)
+from dipy.align.scalespace import IsotropicScaleSpace
+from dipy.core.optimize import SCIPY_LESS_0_12
+
+
+class MattesMIMetric(MattesBase):
+    def __init__(self, nbins=32, sampling_proportion=None):
+        r""" Initializes an instance of the Mattes MI metric
+
+        This class implements the methods required by Optimizer to drive the
+        registration process by making calls to the low level methods defined
+        in MattesBase.
+
+        Parameters
+        ----------
+        nbins : int, optional
+            the number of bins to be used for computing the intensity
+            histograms. The default is 32.
+        sampling_proportion : int in (0,100], optional
```

Comment:
No problem, I will change it to (0,1]. Apparently passing an integer turned out to be confusing (I just try to avoid using floating point variables whenever possible). I also updated the docstring in an attempt to clarify the difference between sparse and dense sampling. Note that if we know that all points will be used, we don't need to actually create an array with the coordinates of the selected points, we can just iterate over all the volume, but if we choose to do so, then we cannot add the random movement, otherwise we would need to actually generate the array of coordinates. So, if we want dense sampling plus random movement, we just need to specify 100% `sampling_proportion` instead of `None`